src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -290,16 +294,61 @@ void WriteTypeMappings (List<TypeDefinition> types)
 			using (var gen = UseSharedRuntime
 				? new TypeNameMapGenerator (types, Log.LogDebugMessage)
 			        : new TypeNameMapGenerator (ResolvedAssemblies.Select (p => p.ItemSpec), Log.LogDebugMessage)) {
-				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.jm"), gen.WriteJavaToManaged);
-				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.mj"), gen.WriteManagedToJava);
+				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.jm"), "jm", gen.WriteJavaToManaged);
+				UpdateWhenChanged (Path.Combine (OutputDirectory, "typemap.mj"), "mj", gen.WriteManagedToJava);
 			}
 		}
 
-		void UpdateWhenChanged (string path, Action<Stream> generator)
+		void UpdateWhenChanged (string path, string type, Action<Stream> generator)
 		{
-			using (var stream = new MemoryStream ()) {
+			if (!EmbedAssemblies) {
+				using (var stream = new MemoryStream ()) {
+					generator (stream);
+					MonoAndroidHelper.CopyIfStreamChanged (stream, path);
+				}
+			}
+
+			string dataFilePath = $"{path}.inc";
+			using (var stream = new NativeAssemblyDataStream ()) {
 				generator (stream);
-				MonoAndroidHelper.CopyIfStreamChanged (stream, path);
+				stream.Flush ();
+				MonoAndroidHelper.CopyIfStreamChanged (stream, dataFilePath);
+
+				var generatedFiles = new List <ITaskItem> ();
+				string mappingFieldName = $"{type}_typemap";
+				string dataFileName = Path.GetFileName (dataFilePath);
+				NativeAssemblerTargetProvider asmTargetProvider;
+				foreach (string abi in SupportedAbis.Split (new char[] { ',', ';' }, StringSplitOptions.RemoveEmptyEntries)) {
+					switch (abi.Trim ()) {
+						case "armeabi-v7a":
+							asmTargetProvider = new ARMNativeAssemblerTargetProvider (is64Bit: false);
+							break;
+
+						case "arm64-v8a":
+							asmTargetProvider = new ARMNativeAssemblerTargetProvider (is64Bit: true);
+							break;
+
+						case "x86":
+							asmTargetProvider = new X86NativeAssemblerTargetProvider (is64Bit: false);
+							break;
+
+						case "x86_64":
+							asmTargetProvider = new X86NativeAssemblerTargetProvider (is64Bit: true);
+							break;
+
+						default:
+							throw new InvalidOperationException ($"Unknown ABI {abi}");
+					}
+
+					var asmgen = new TypeMappingNativeAssemblyGenerator (asmTargetProvider, stream, dataFileName, stream.MapByteCount, mappingFieldName);
+					string asmFileName = $"{path}.{abi.Trim ()}.s";
+					using (var ms = new MemoryStream ()) {
```

Comment:
Can we reuse a single `MemoryStream` here for each abi? You can do `stream.SetLength(0)` and it keeps the underlying buffer reserved and you can write to it again.